### PR TITLE
Add databases and organizations

### PR DESCRIPTION
**Pull request by zugao**
## Summary

This PR rewrites the previous unfinished Database reconciliation and adds Organization reconciler. 

### CRDs
This is an example of a custom resource Database and Organization. The resources are cluster scoped:
```
apiVersion: stardog.vshn.ch/v1beta1
kind: Database
metadata:
  name: db
spec:
  databaseName: db
  namedGraphPrefix: https://graph.db.net/
  options: |
    {
      "transaction.write.conflict.strategy": "abort_on_conflict",
      "index.aggregate":  "On"
    }
  stardogInstanceRefs:
    - name: stardog-instance-test
      namespace: stardog-test
    - name: stardog-instance-prod
      namespace: stardog-prod

apiVersion: stardog.vshn.ch/v1beta1
kind: Organization
metadata:
  name: org-test
spec:
  name: org-test
  displayName: Organization Test
  databaseRef: db
  namedGraphs:
    - graph1
    - graph2
```

### Main operation
Database:
The idea is that the database [spec.databaseName] is created across multiples Stardog instances [spec.stardogInstanceRefs] (Stardog servers) with a specific database configuration [spec.Options]. The graph name [spec.namedGraphPrefix] will be used subsequently by an Organization Reconciler to create full paths to [Named Graphs](https://docs.stardog.com/operating-stardog/security/named-graph-security#:~:text=Stardog%20allows%20a%20set%20of,all%20users%20running%20reasoning%20queries.).
Organization:
The reconciler adds fine grained permissions for the organization.

### Behind the scenes
Creating the database is only a tiny part of this reconciler. These are the added resources that follow after a database creation:
* 2 users in Stardog (read and write)
* 1 secret in k8s with credentials of above users set in the same namespace as as StardogInstace resource
* 2 roles in Stardog (read and write)
* 4 types of permissions in Stardog
* 3 role assignments to the users in Stardog.

Creating an organization the following resources are created:
* 1 user in Stardog
* 1 secret in k8s with credentials of above user set in the same namespace as as StardogInstace resource
* 1 role in Stardog
* 2 permissions to access read/write the database in Stardog
* permissions to access read/write specific paths (named graphs) base on the number of graphs in Stardog.
* role assignments to the users in Stardog.

### Requirements
* Only the number of instances are editable in the Database resource.
* A Database CRD cannot be deleted unless databases for each instance is empty and there are no Organization CRDs assigned to it.
* An instance from a Database resource can be removed only if the database is empty.
* Successfully deleting an instance from a Database resource will also trigger the removal of the organization resources from that instance. The Organization CRD will remain but the status will be updated to how many database instances it is still connected.

## Checklist

- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog

### CODE REVIEW

File: controllers/database_controller.go [hunk 1]
```diff
@@ -38,153 +62,405 @@ func (r *DatabaseReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
+
+	if !slices.Contains(writeUserRolesResp.Payload.Roles, writeName) {
+		params := users_roles.NewAddRoleParams().
+			WithUser(writeName).
+			WithRole(&models.Rolename{Rolename: &writeName})
+		roleResp, err := stardogClient.UsersRoles.AddRole(params, auth)
+		if err != nil || !roleResp.IsSuccess() {
+			r.Log.Error(err, "error assigning role to user", "user", writeName, "role", writeName)
+			return err
 		}
```
Comment (glrf) [inline controllers/database_controller.go:451]: Nitpicky Nitpick (feel free to ignore): I'd probably move the logic to create user roles to a helper function `assignDefaultRoles` or something. The same way you did for the permissions. 

You're mixing layers here which makes it a bit harder to read.

File: controllers/organization_controller.go [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+
+		err := r.Client.Status().Update(ctx, organization)
+		if err != nil {
+			r.Log.Error(err, "Could not update Organization Status", getLoggingKeysAndValuesForOrganization(organization)...)
+			return err
+		}
+	}
+
+	// If status is set for database name then we treat it as an update (2 - n object reconciliation)
+	if status.Name != "" {
```
Comment (glrf) [inline controllers/organization_controller.go:166]: At that point `status.Name` is always set

**Review by glrf — APPROVED** [review]
I think this generally looks OK to me, I just have some minor nitpicks, but I have no idea how stardog works,  so I might have missed something.

One general nitpick is that I don't like all the `*Reconciliation` and `ReconciliationContext`. That IMO just promotes lazily passing all the things everywhere. But as far as I can tell this was already part of the codebase and I don't see much value in fixing this right now.

File: controllers/database_controller.go [hunk 3]
```diff
@@ -38,153 +62,405 @@ func (r *DatabaseReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
+		return ctrl.Result{Requeue: false}, nil
 	}
 
-	// Create Stardog resources
-	instance := &stardogv1beta1.Instance{}
-	err = r.Get(ctx, types.NamespacedName{Name: database.Spec.InstanceRef.Name, Namespace: database.Namespace}, instance)
+	if err := r.validateSpecification(dr.reconciliationContext.context, dr.resource); err != nil {
+		rc.SetStatusCondition(createStatusConditionInvalid(err))
+		rc.SetStatusCondition(createStatusConditionReady(false, "Specification cannot be validated"))
+		return ctrl.Result{Requeue: false}, r.updateStatus(dr)
```
[the remark this answers is not in the document]
Comment (zugao) [reply controllers/database_controller.go]: Added the logs everywhere.

File: controllers/database_controller.go [hunk 4]
```diff
@@ -38,153 +62,405 @@ func (r *DatabaseReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
-		r.Log.Info("created Stardog database", "name", database.Spec.DatabaseName)
 	}
 
-	// Get user credentials from Secret
-	secret := &corev1.Secret{}
-	err = r.Get(ctx, types.NamespacedName{Name: secretName, Namespace: database.Namespace}, secret)
-	if err != nil {
-		return ctrl.Result{}, err
+	// If status is set for database name then we treat it as an update (2 - n object reconciliation)
+	if status.DatabaseName != "" {
```
[the remark this answers is not in the document]
Comment (zugao) [reply controllers/database_controller.go:310]: No It will, I just checked the code and it will always overwrite the spec for those 3 attributes.

**Comment by zugao** [comment]
> One general nitpick is that I don't like all the *Reconciliation and ReconciliationContext. That IMO just promotes lazily passing all the things everywhere. But as far as I can tell this was already part of the codebase and I don't see much value in fixing this right now.

Yes this was part of the codebase and I agree that it would be nice to remove these 2 big structs and create more refined types. Also having more files to better abstract the logic would be also welcomed, especially when there're a lot of places for code reuse.

**Pull request merged**